### PR TITLE
TC Automation Issue multiple custom restores at the same time using different storage class mapping

### PR DESCRIPTION
**Pull request by sn-px**
**What this PR does / why we need it**:
TC Automation Issue multiple custom restores at the same time using different storage class mapping

**Which issue(s) this PR fixes** (optional)
Closes #PA-791

**Special notes for your reviewer**:
Latest log :
https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/1000/consoleFull

http://aetos.pwx.purestorage.com/resultSet/testSetID/151106

### CODE REVIEW

File: tests/backup/backup_restore_basic_test.go [hunk 1]
```diff
@@ -1318,3 +1322,159 @@ var _ = Describe("{BackupSyncBasicTest}", func() {
...
+		})
+		Step("Restoring multiple backup in different storage class in same cluster", func() {
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			pvcs, err := core.Instance().GetPersistentVolumeClaims(bkpNamespaces[0], labelSelectors)
+			singlePvc := pvcs.Items[0]
+			sourceScName, err = core.Instance().GetStorageClassForPVC(&singlePvc)
+			var wg sync.WaitGroup
+			for _, scName := range scNames {
+				log.InfoD("Restoring backup in different storage class  in same cluster")
```
Comment (snigdha-px) [inline tests/backup/backup_restore_basic_test.go:1437]: Remove 2 spaces between class  in

File: tests/backup/backup_restore_basic_test.go [hunk 2]
```diff
@@ -1318,3 +1322,159 @@ var _ = Describe("{BackupSyncBasicTest}", func() {
...
+				backupLocationUID = uuid.New()
+				backupLocationMap[backupLocationUID] = bkpLocationName
+				err := CreateCloudCredential(provider, cloudCredName, cloudCredUID, orgID, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of cloud credential named [%s] for org [%s] with [%s] as provider", cloudCredName, orgID, provider))
+				err = CreateBackupLocation(provider, bkpLocationName, backupLocationUID, cloudCredName, cloudCredUID, getGlobalBucketName(provider), orgID, "")
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Creating backup location %s", bkpLocationName))
+			}
+		})
+		Step("Taking backup of application for different combination of restores", func() {
+			log.InfoD("Taking  backup of application for different combination of restores")
```
Comment (snigdha-px) [inline tests/backup/backup_restore_basic_test.go:1423]: 2 spaces between Taking  backup

**Review by snigdha-px — COMMENTED** [review]
minor change

File: tests/backup/backup_restore_basic_test.go [hunk 3]
```diff
@@ -1318,3 +1322,159 @@ var _ = Describe("{BackupSyncBasicTest}", func() {
...
+		Step("Register cluster for backup", func() {
+			err = CreateSourceAndDestClusters(orgID, "", "", ctx)
+			dash.VerifyFatal(err, nil, "Creating source and destination cluster")
+			clusterStatus, err = Inst().Backup.GetClusterStatus(orgID, SourceClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", SourceClusterName))
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", SourceClusterName))
+			clusterUid, err = Inst().Backup.GetClusterUID(ctx, orgID, SourceClusterName)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching [%s] cluster uid", SourceClusterName))
+		})
+		Step("Create new storage class on source and destination cluster for storage class mapping for restore", func() {
```
Comment (snigdha-px) [inline tests/backup/backup_restore_basic_test.go:1384]: Here storage class is being created only on source cluster,right?

Comment (snigdha-px) [reply tests/backup/backup_restore_basic_test.go:1384]: But in Step it says Create new storage class on source and destination cluster

Comment (sn-px) [reply tests/backup/backup_restore_basic_test.go]: thanks ..corrected it

File: tests/backup/backup_restore_basic_test.go [hunk 4]
```diff
@@ -1318,3 +1322,159 @@ var _ = Describe("{BackupSyncBasicTest}", func() {
...
+			}
+		})
+		Step("Taking backup of application for different combination of restores", func() {
+			log.InfoD("Taking backup of application for different combination of restores")
+			backupName = fmt.Sprintf("%s-%s-%v", BackupNamePrefix, bkpNamespaces[0], time.Now().Unix())
+			err = CreateBackup(backupName, SourceClusterName, bkpLocationName, backupLocationUID, []string{bkpNamespaces[0]},
+				labelSelectors, orgID, clusterUid, "", "", "", "", ctx)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Taking backup: %s", backupName))
+		})
+		Step("Restoring multiple backup in different storage class in same cluster", func() {
```
Comment (snigdha-px) [inline tests/backup/backup_restore_basic_test.go:1429]: nit: Multiple restore for same backup in different storage class in same cluster

**Review by snigdha-px — APPROVED** [review]
LGTM

**Review by ak-px — APPROVED** [review]
LGTM

File: tests/backup/backup_restore_basic_test.go (hunk 4 above)
Comment (sn-px) [reply tests/backup/backup_restore_basic_test.go]: done

File: tests/backup/backup_restore_basic_test.go [hunk 5]
```diff
@@ -1442,3 +1446,159 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+	namespaceMap := make(map[string]string)
+	backupLocationMap := make(map[string]string)
+	labelSelectors := make(map[string]string)
+	bkpNamespaces = make([]string, 0)
+	storageClassMapping := make(map[string]string)
+	k8sStorage := storage.Instance()
+	params := make(map[string]string)
+
+	JustBeforeEach(func() {
+		StartTorpedoTest("IssueMultipleCustomRestoreAtSameTimeWithDiffStorageClassMapping",
```
Comment (mkoppal-px) [inline tests/backup/backup_restore_basic_test.go:1476]: Change the string to `MultipleCustomRestoreSameTimeDiffStorageClassMapping` to keep it consistent with the `Describe` string

File: tests/backup/backup_restore_basic_test.go [hunk 6]
```diff
@@ -1442,3 +1446,159 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+			dash.VerifyFatal(err, nil, "Creating source and destination cluster")
+			clusterStatus, err = Inst().Backup.GetClusterStatus(orgID, SourceClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", SourceClusterName))
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", SourceClusterName))
+			clusterUid, err = Inst().Backup.GetClusterUID(ctx, orgID, SourceClusterName)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching [%s] cluster uid", SourceClusterName))
+		})
+		Step("Create new storage class on source cluster for storage class mapping for restore", func() {
+			log.InfoD("Create new storage class on source cluster for storage class mapping for restore")
+			scCount := 2
```
Comment (mkoppal-px) [inline tests/backup/backup_restore_basic_test.go:1510]: Declare this at global level under Describe

**Comment by mkoppal-px** [comment]
Why are we doing only 2 restores in parallel? Can we not do like 10 restores at least?
@kshithijiyer-px Need your input here

**Comment by kshithijiyer-px** [comment]
> Why are we doing only 2 restores in parallel? Can we not do like 10 restores at least? @kshithijiyer-px Need your input here

I am not sure if 10 restores can go through on the same cluster in parallel due to size of the cluster would I would suggest to do is try out but make this configureable so that we can alter this parameter and run is with as much load as we want.

File: tests/backup/backup_restore_basic_test.go [hunk 7]
```diff
@@ -1442,3 +1446,159 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+		})
+		Step("Taking backup of application for different combination of restores", func() {
+			log.InfoD("Taking backup of application for different combination of restores")
+			backupName = fmt.Sprintf("%s-%s-%v", BackupNamePrefix, bkpNamespaces[0], time.Now().Unix())
+			err = CreateBackup(backupName, SourceClusterName, bkpLocationName, backupLocationUID, []string{bkpNamespaces[0]},
+				labelSelectors, orgID, clusterUid, "", "", "", "", ctx)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Taking backup: %s", backupName))
+		})
+		Step("Multiple restore for same backup in different storage class in same cluster at the same time", func() {
+			ctx, err := backup.GetAdminCtxFromSecret()
```
[the remark this answers is not in the document]
Comment (sn-px) [reply tests/backup/backup_restore_basic_test.go:1556]: addded

File: tests/backup/backup_restore_basic_test.go (hunk 6 above)
Comment (sn-px) [reply tests/backup/backup_restore_basic_test.go]: done

**Comment by sn-px** [comment]
> Why are we doing only 2 restores in parallel? Can we not do like 10 restores at least? @kshithijiyer-px Need your input here

Increased the parallel restore count to 10

File: tests/backup/backup_restore_basic_test.go [hunk 8]
```diff
@@ -1442,3 +1446,161 @@ var _ = Describe("{BackupMultipleNsWithSameLabel}", func() {
...
+		Step("Multiple restore for same backup in different storage class in same cluster at the same time", func() {
+			log.InfoD(fmt.Sprintf("Multiple restore for same backup into %d different storage class in same cluster at the same time", scCount))
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			pvcs, err := core.Instance().GetPersistentVolumeClaims(bkpNamespaces[0], labelSelectors)
+			singlePvc := pvcs.Items[0]
+			sourceScName, err = core.Instance().GetStorageClassForPVC(&singlePvc)
+			var wg sync.WaitGroup
+			for _, scName := range scNames {
+				log.InfoD("Restoring backup in different storage class in same cluster at the same time")
```
Comment (mkoppal-px) [inline tests/backup/backup_restore_basic_test.go:1563]: Do not put this in the loop. Keep it outside.

**Review by mkoppal-px — APPROVED** [review]
LGTM except for a minor comment

**Pull request merged**